Add MANIFEST.in

malfet opened the pull request:
Do not build PyTorch if `setup.py` is called with  'sdist' option
Regenerate bundled license while sdist package is being built
Refactor `check_submodules` out of `build_deps` and check that submodules project are present during source package build stage.

Test that sdist package is configurable during `asan-build` step

Fixes #52843

malfet committed:
Add MANIFEST.in

Also, do not build PyTorch when 'sdist' option is called

mattip commented:
In order to correctly concatentate the two license files `./LICENSE` and `./third_party/LICENSES_BUNDLED.txt`, the `sdist` command should be wrapped like the `class wheel_concatenate(bdist_wheel)` class.

rgommers commented on MANIFEST.in:
```diff
@@ -0,0 +1,19 @@
+recursive-include scripts *.*
+recursive-include mypy_plugins *.*
+recursive-include modules *.*
+global-exclude *.o *.so *.dylib *.a .git
```
You'll probably want to ensure to not do this in a dirty tree.

E.g. what SciPy does: https://github.com/scipy/scipy/blob/master/setup.py#L202

- in the `sdist` command in `setup.py`, check if all submodules are up to date.
- In the release script, ensure a `git clean -xdf` or running it on a clean git clone.

malfet replied:
Agree about the clean, but not necessarily about keeping submodules uptodate, as one should be at free to package their repo in whatever state of dirtiness into their `sdist` and `bdist` packages.

rgommers replied:
You can always just commit whatever you want to package on a branch (even a temporary one), right? I'm not sure what the use case is for packaging things that aren't  committed anywhere.

Up to you of course, I just know SciPy implemented it after shipping an incomplete release and we've never had a complaint about the check.

malfet replied:
I've extracted existing submodules validation logic and share it across 'sdist' and 'build_deps' steps, will add some of your suggestions in the followup up diff.

xuzhao9 commented on MANIFEST.in:
```diff
@@ -0,0 +1,19 @@
+recursive-include scripts *.*
+recursive-include mypy_plugins *.*
+recursive-include modules *.*
+global-exclude *.o *.so *.dylib *.a .git
```
Maybe we should add more global-exclude file patterns similar to SciPy, such as `*.swp`, `*.pyc`? https://github.com/scipy/scipy/blob/master/MANIFEST.in#L31

malfet committed:
Partially Incorporate review feedback

Add MANIFEST to .gitignore files
Add few more extensions to ignore/pruning list in MANIFEST.in
Update bundled licenses during sdist installation
Extract `check_submodules` routine from `build_deps`

malfet committed:
Test that sdist package is configurable

mattip commented on .jenkins/pytorch/build.sh:
```diff
@@ -317,3 +317,15 @@ if [[ "${BUILD_ENVIRONMENT}" == *xla* ]]; then
   assert_git_not_dirty
 fi
+
+# Test that source package is configurable
```
```suggestion
# Test building via the sdist source tarball
```

mattip commented:
I expected to see some traces of the build in the [pytorch_linux_xenial_py3_clang5_asan_build](https://app.circleci.com/pipelines/github/pytorch/pytorch/278769/workflows/3a9cc440-3df6-4e72-90ee-4e8e09af21d0/jobs/11193255/parallel-runs/0/steps/0-106), but it seems that job calls [`exec .. build-asan.sh`](https://github.com/pytorch/pytorch/blob/v1.8.0-rc1/.jenkins/pytorch/build.sh#L15) and exits without continuing. Is there another `*asan*` build that hits the new test?

malfet committed:
Fix typo

malfet committed:
Applied @mattip suggestion

malfet commented on setup.py:
```diff
@@ -623,8 +628,14 @@ def run(self):
 
-        # It's an old-style class in Python 2.7...
-        distutils.command.clean.clean.run(self)
+        super().run(self)
```
Fixed typo
```suggestion
        super().run()
```

malfet committed:
Move test sdist logic from build.sh to build-asan.sh

malfet commented:
> I expected to see some traces of the build in the [pytorch_linux_xenial_py3_clang5_asan_build](https://app.circleci.com/pipelines/github/pytorch/pytorch/278769/workflows/3a9cc440-3df6-4e72-90ee-4e8e09af21d0/jobs/11193255/parallel-runs/0/steps/0-106), but it seems that job calls [`exec .. build-asan.sh`](https://github.com/pytorch/pytorch/blob/v1.8.0-rc1/.jenkins/pytorch/build.sh#L15) and exits without continuing. Is there another `*asan*` build that hits the new test?

Thank you for spotting. Moved the test logic from `build.sh` to `build-asan.sh`

malfet committed:
Slightly change test script

mattip commented:
Cool. 
The [build log](https://app.circleci.com/pipelines/github/pytorch/pytorch/279654/workflows/e80fd11f-1ad1-46ed-a8ed-49cf929b84ac/jobs/11235905) has a `warning: sdist: standard file not found: should have one of README, README.txt` at the beginning of `python setup.py sdist`, and lots and lots of lines about creating the sdist. 

Then at the beginning of the cmake-only build (`python setup.py build --cmake-only`) there is a `fatal: Not a git repository (or any of the parent directories): .git` but it is not really fatal. Other than that, the cmake warnings seem to be the usual ones I see when building from a git checkout.
